### DISCUSSION TITLE
static image support

### DISCUSSION THREAD

**Discussion by gaultierq** [ElMassimo/vite-plugin-image-presets#21] [Q&A] [answered]
Hello,
I am using vite in a ruby on rails project, using the `vite-plugin-ruby` plugin. 
I am now trying to normalize my static images - not referenced by any javascript - during build time. 

I stumbled across the `vite-plugin-image-preset`, and I was wondering if it could work with the `vite-plugin-ruby`. 
My understanding is that the `vite-plugin-ruby` is "just" registering assets with a hash. 
Is there a way I could override this behavior and leverage what is done in the `vite-plugin-image-preset` plugin ?
Or maybe add a step generating a javascript file referencing all my static images ?

Thanks for all your work in the vite ecosystem

**Comment by ElMassimo** [comment, marked as answer]
Hi Quentin!

This plugin will ["generate JS" for imports of images](https://github.com/ElMassimo/vite-plugin-image-presets/blob/main/src/index.ts#L53).

In contrast, `vite-plugin-ruby` will [copy files directly and generate a `manifest-assets.json` file](https://github.com/ElMassimo/vite_ruby/blob/main/vite-plugin-ruby/src/manifest.ts#L33-L50) to compensate for Vite's lack of support for adding non-(JS or CSS) assets as entrypoints.

As a result, there's no direct way to integrate both plugins.

---

> I am now trying to normalize my static images 

You could create a `post` Vite plugin with a [`writeBundle`](https://rollupjs.org/plugin-development/#writebundle) callback, walk through `config.build.assetsDir` and manually use `sharp` to optimize assets in place.

A disadvantage of optimizing in-place is that the fingerprint would not change, so changing your optimization params wouldn't affect the result, but if you are doing standard optimizations maybe that's acceptable (and would be a very easy way to achieve optimally sized images).

**Reply by gaultierq** [reply]
Thanks for your answer, it helped wrap my head around it. I will give it a try !

**Reply by gaultierq** [reply]
Hey @ElMassimo  
I have started playing around with the vite plugin api. I went to your `vite-plugin-ruby` for inspiration. 
So far, I managed to produce resized images during the `generateBundle` step, and  emit the files with a custom suffix. 

However, adding these file to the `manifest-assets.json` will override the manifest produced by the ruby plugin. 

I was wondering if merging the 2 manifest could be possible, or even a good idea. 

If so, do you know how to achieve that ? The file are not existing yet, and I don't know how to get my hand on the `manifest-asset.json` reference id.

Many thanks

**Reply by ElMassimo** [reply]
Try adding a [`writeBundle`](https://rollupjs.org/plugin-development/#writebundle) hook in your plugin:

> Called only at the end of bundle.write() once all files have been written.

---

Merging manifests is something that might be harder to debug, so you _could_ instead redefine [`manifest_paths`](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/vite_ruby/config.rb#L31-L33) in `ViteRuby::Config` :

```ruby
ViteRuby::Config.prepend Module.new {
  # Override: Add manifest for images.
  def manifest_paths
    super + [build_output_dir.join('manifest-images.json')]
  end
}
```

When overriding internal methods you risk breakage on upgrades, as the library could remove or rename them. You'll want to assess whether this risk is acceptable for your use case.
